navigator_engine/common/decision_engine.py

```diff
@@ -0,0 +1,115 @@
+import ast
+import navigator_engine.model as model
+from navigator_engine.common import CONDITIONAL_FUNCTIONS, DATA_LOADERS, DecisionError
+from navigator_engine.common.progress_tracker import ProgressTracker
+from typing import Callable
+
+
+class DecisionEngine():
+
+    def __init__(self, graph: model.Graph, data: object,
+                 skip: list[str] = [], route: list[str] = []) -> None:
+        self.graph = graph
+        self.network = self.graph.to_networkx()
+        self.data = data
+        self.skip = skip
+        self.root_node = self.get_root()
+        self.skipped = []
+        self.progress = ProgressTracker(self.network, route=route)
+
+    def decide(self, data: object = None, skip: list[str] = None) -> dict:
+        if data:
+            self.data = data
+        if skip:
+            self.skip = skip
+        self.progress.reset()
+        self.skipped = []
+        self.decision = self.process_node(self.root_node)
+        return self.decision
+
+    def get_root(self) -> model.Node:
+        for node, in_degree in self.network.in_degree():
+            if in_degree == 0:
+                return node
+        raise DecisionError("Graph {graph.id} ({graph.title}) has no root node")
```

Comment:
That probably needs an `f"` to be a f-string.

Reply:
Good catch